File: apps/scala/design-pattern/src/Builder.scala
// Builder パターン
// case class の copy メソッドとフルーエント API でオブジェクトを段階的に構築する

package designpattern.builder

case class Computer(
  cpu: String = "unknown",
  memory: Int = 0,
  storage: Int = 0,
  gpu: Option[String] = None,
  display: Option[String] = None
):
  def describe: String =
    val parts = List(
      s"CPU: $cpu",
      s"メモリ: ${memory}GB",
      s"ストレージ: ${storage}GB"
    ) ++ gpu.map(g => s"GPU: $g").toList ++
      display.map(d => s"ディスプレイ: $d").toList
    parts.mkString(", ")

class ComputerBuilder:
  private var computer: Computer = Computer()

  def setCpu(cpu: String): ComputerBuilder =
    computer = computer.copy(cpu = cpu)
    this

  def setMemory(memory: Int): ComputerBuilder =
    computer = computer.copy(memory = memory)
    this

  def setStorage(storage: Int): ComputerBuilder =
    computer = computer.copy(storage = storage)
    this

  def setGpu(gpu: String): ComputerBuilder =
    computer = computer.copy(gpu = Some(gpu))
    this

  def setDisplay(display: String): ComputerBuilder =
    computer = computer.copy(display = Some(display))
    this

  def build(): Computer = computer

// ディレクタ: 定型構成を提供する
object ComputerDirector:
  def gamingComputer: Computer =
    ComputerBuilder()
      .setCpu("Intel i9")
      .setMemory(32)
      .setStorage(1000)
      .setGpu("RTX 4090")
      .setDisplay("27インチ 4K")
      .build()

  def officeComputer: Computer =
    ComputerBuilder()
      .setCpu("Intel i5")
      .setMemory(16)
      .setStorage(512)
      .build()

  def laptopComputer: Computer =
    ComputerBuilder()
      .setCpu("Apple M3")
      .setMemory(16)
      .setStorage(256)
      .setDisplay("14インチ Retina")
      .build()

File: apps/haskell/design-pattern/src/Builder.hs
-- | Builder パターン
-- レコード構文のデフォルト値 + 更新関数で段階的に構築する。
-- Haskell のレコード更新構文がビルダーの役割を果たす。
module Builder
  ( Computer(..)
  , defaultComputer
  , setDisplay
  , setMemory
  , setStorage
  , setOS
  , buildGaming
  , buildOffice
  , buildServer
  , computerSummary
  ) where

-- | コンピュータ
data Computer = Computer
  { cpuType    :: String
  , display    :: String
  , memory     :: Int     -- ^ GB
  , storage    :: Int     -- ^ GB
  , os         :: String
  , gpu        :: String
  } deriving (Show, Eq)

-- | デフォルトのコンピュータ
defaultComputer :: Computer
defaultComputer = Computer
  { cpuType = "Intel i5"
  , display = "15.6 inch"
  , memory  = 8
  , storage = 256
  , os      = "Linux"
  , gpu     = "Integrated"
  }

-- | ディスプレイを設定
setDisplay :: String -> Computer -> Computer
setDisplay d c = c { display = d }

-- | メモリを設定
setMemory :: Int -> Computer -> Computer
setMemory m c = c { memory = m }

-- | ストレージを設定
setStorage :: Int -> Computer -> Computer
setStorage s c = c { storage = s }

-- | OS を設定
setOS :: String -> Computer -> Computer
setOS o c = c { os = o }

-- | ゲーミング PC を構築
buildGaming :: Computer
buildGaming = defaultComputer
  { cpuType = "Intel i9"
  , display = "27 inch 4K"
  , memory  = 64
  , storage = 2000
  , os      = "Windows"
  , gpu     = "NVIDIA RTX 4090"
  }

-- | オフィス PC を構築
buildOffice :: Computer
buildOffice = defaultComputer
  { cpuType = "Intel i3"
  , display = "24 inch"
  , memory  = 16
  , storage = 512
  , os      = "Windows"
  , gpu     = "Integrated"
  }

-- | サーバを構築
buildServer :: Computer
buildServer = defaultComputer
  { cpuType = "AMD EPYC"
  , display = "None"
  , memory  = 256
  , storage = 8000
  , os      = "Linux"
  , gpu     = "None"
  }

-- | コンピュータのサマリ
computerSummary :: Computer -> String
computerSummary c =
  cpuType c ++ " / " ++ display c
  ++ " / " ++ show (memory c) ++ "GB RAM"
  ++ " / " ++ show (storage c) ++ "GB SSD"
  ++ " / " ++ os c
  ++ " / GPU: " ++ gpu c
